sycl/source/detail/platform_impl.cpp

```diff
@@ -121,95 +122,172 @@ vector_class<platform> platform_impl::get_platforms() {
+      }
+      decDescs.back().platName = allowList.substr(start, pos - start);
+      pos = pos + 2;
 
-    *size = str - *valuePtr;
+      if (allowList[pos] == ',') {
+        pos++;
+      }
 
-    // Skip closing sequence "}}"
-    str += 2;
+    }
 
-    if ('\0' == *str)
-      break;
+    else if ((allowList.compare(pos, platformVersion.size(),
+                                platformVersion)) == 0) {
+      prev = pos;
+      if ((pos = allowList.find("{{", pos)) == std::string::npos) {
+        throw sycl::runtime_error("Malformed syntax in SYCL_DEVICE_ALLOWLIST",
+                                  PI_INVALID_VALUE);
+      }
+      if (pos > prev + platformVersion.size()) {
+        throw sycl::runtime_error("Malformed syntax in SYCL_DEVICE_ALLOWLIST",
+                                  PI_INVALID_VALUE);
+      }
 
-    // '|' means that the is another filter
-    if ('|' == *str)
+      size_t start = pos + 2;
+      if ((pos = allowList.find("}}", pos)) == std::string::npos) {
+        throw sycl::runtime_error("Malformed syntax in SYCL_DEVICE_ALLOWLIST",
+                                  PI_INVALID_VALUE);
+      }
+      decDescs.back().platVer = allowList.substr(start, pos - start);
+      pos = pos + 2;
+    }
+
+    else if (allowList.find('|', pos) != std::string::npos) {
+      pos = allowList.find('|') + 1;
+      while (allowList[pos] == ' ') {
+        pos++;
+      }
       decDescs.emplace_back();
-    else if (',' != *str)
-      throw sycl::runtime_error("Malformed device allowlist", PI_INVALID_VALUE);
+    }
 
-    ++str;
+    else {
+      throw sycl::runtime_error("Unrecognized key in device allowlist",
+                                PI_INVALID_VALUE);
+    }
+  } // while (pos <= allowList.size())
+  return decDescs;
+}
+
+std::vector<int> convertVersionString(std::string version) {
+  // version string format is xx.yy.zzzzz
+  std::vector<int> values;
+  size_t pos = 0;
+  size_t start = pos;
+  if ((pos = version.find(".", pos)) == std::string::npos) {
+    throw sycl::runtime_error("Malformed syntax in version string",
+                              PI_INVALID_VALUE);
+  }
+  values.push_back(std::stoi(version.substr(start, pos)));
+  pos++;
+  start = pos;
+  if ((pos = version.find(".", pos)) == std::string::npos) {
+    throw sycl::runtime_error("Malformed syntax in version string",
+                              PI_INVALID_VALUE);
   }
+  values.push_back(std::stoi(version.substr(start, pos)));
+  pos++;
+  values.push_back(std::stoi(version.substr(pos)));
 
-  return decDescs;
+  return values;
+}
+
+enum MatchState { UNKNOWN, MATCH, NOMATCH };
+
+MatchState matchVersions(std::string version1, std::string version2) {
+  std::vector<int> v1 = convertVersionString(version1);
+  std::vector<int> v2 = convertVersionString(version2);
+  if (v1[0] >= v2[0] && v1[1] >= v2[1] && v1[2] >= v2[2]) {
+    return MatchState::MATCH;
+  } else {
+    return MatchState::NOMATCH;
+  }
 }
```

Comment:
I believe it's an implementation of "greater or equal" comparision `Version1 >= Version2`.
The current condition is only sufficient to say that `V1 >= V2` but it lacks needed one.
Version is a number which could be thought of as noted with variable base and should be compared like this:
```
  V1 >= V2 iff V1.M > V2.M || (V1.M == V2.M && (V1.m > V2.m || (V1.m == V2.m && V1.p >= V2.p)))
  Vx.M - major
  Vx.m - minor
  Vx.p - patch
```

Reply:
Thanks for the catch.  It is better now.